db/migrate/20181128054803_simplify_variant_override_on_demand_and_count_on_hand.rb

```diff
@@ -0,0 +1,29 @@
+# This simplifies variant overrides to have only the following combinations:
+#
+#    on_demand | count_on_hand
+#   -----------+---------------
+#    true      | nil
+#    false     | set
+#    nil       | nil
+#
+# Refer to the table {here}[https://github.com/openfoodfoundation/openfoodnetwork/issues/3067] for
+# the effect of different variant and variant override stock configurations.
+#
+# Furthermore, this will allow all existing variant overrides to satisfy the newly added model
+# validation rules.
+class SimplifyVariantOverrideOnDemandAndCountOnHand < ActiveRecord::Migration
+  def up
+    # When on_demand is nil but count_on_hand is set, force limited stock.
+    VariantOverride.where(on_demand: nil).where("count_on_hand IS NOT NULL")
+      .update_all(on_demand: false)
+
+    # Clear count_on_hand if forcing on demand.
+    VariantOverride.where(on_demand: true).update_all(count_on_hand: nil)
+
+    # When on_demand is false but count on hand is not specified, set this to use producer stock
+    # settings.
+    VariantOverride.where(on_demand: false, count_on_hand: nil).update_all(on_demand: nil)
+  end
+
+  def down; end
```

Comment:
@mkllnk No, I haven't seen that yet. :+1: 

> Creating records for all affected rows would be good in case there are unknown side effects.

You are talking about creating a DB table for caching the original values, right? Yes, this would allow undoing the changes, which is safer.

Regarding naming of this table, I can shorten the DB migration to "SimplifyVariantOverrideStockSettings", and then name the table "simplify_variant_override_stock_settings_changed_variant_overrides" with `variant_override_id`, `on_demand`, and `count_on_hand` columns. Do let me know if this is what you had in mind.

Thank you!

Reply:
> You are talking about creating a DB table for caching the original values, right? Yes, this would allow undoing the changes, which is safer.

Isn't that a bit too much? I'm afraid something like this would cost us way too much on dev hours given the size of the change.

Reply:
No, I didn't mean creating another DB table. I meant creating a backup file with the affected records as in the example.

I do agree that it is a considerable overhead and added complexity. We shouldn't do that with every migration. I'm not sure about this one though. Yes, it is simple, but it potentially affects stock levels.

```rb
    VariantOverride.where(on_demand: nil).where("count_on_hand IS NOT NULL")
      .update_all(on_demand: false)
```

So if `on_demand` wasn't set, the variant's on_demand is considered. The stock logic may find that the variant is on demand and ignores the stock level of the override. So this would change the current logic, right?

```rb
VariantOverride.where(on_demand: false, count_on_hand: nil).update_all(on_demand: nil)
```

We are fixing invalid setups here, but they could be in use -- successfully. If the variant has `on_demand: false`, nothing changed. But if the variant is set `on_demand: true`, the stock level is not in use by the producer, but it is used by the overriding shop. They would lose their stock level. The shop may be sharing the stock level with another shop. I would doubt that they set it up intentionally like that. So we can probably copy the stock level from the variant to the override. But if they were sharing the stock level intentionally, this pull request would break their usage scenario.

We should query all production servers for these edge cases. Maybe we need to look at the affected enterprises and tell them about the change.

Changing stock levels is a big deal for enterprises. We had some drama with this recently: https://github.com/openfoodfoundation/openfoodnetwork/pull/2963
Following that incident I had the idea of backup files that allow the rollback of destructive migrations.

Reply:
Ah, I thought you were talking about providing an undo action for the DB migration.

Actually, based on my analysis in the issue description of #3067, the behaviour would not change, with the exception of this scenario:

* VO - OD: `false`, COH: `nil`
* Variant (producer) - OD: `true`, COH: filled

Here, instead of using the COH of the producer - which doesn't seem right because COH is not even visible in the product list, variant list, and variant form when the variant is "On Demand" - the DB migration sets it to use the producer's stock settings. (Noting that this quirk is the root of the problem in #3058.)

Aside from this scenario, do you see any other situation when the behaviour would change?

I agree that we should keep a backup file. So a CSV file with the data, so we can send to instance owners if needed.

I will put in the following data:

* Distributor Name
* Distributor ID
* Product Name
* Product ID
* Variant Description
* Variant ID
* Previous On Demand
* Previous Count on Hand
* Updated On Demand
* Updated Count on Hand

I'm already working on this, but let me know if you find issues.

Reply:
This is the script I used to generate the table in #3067:

```ruby
variant = Spree::Variant.new
variant_override = VariantOverride.new

def set_stock(object, on_demand, count_on_hand)
  object.on_demand = on_demand
  object.count_on_hand = count_on_hand
end

variant.extend(OpenFoodNetwork::ScopeVariantToHub::ScopeVariantToHub)
variant.instance_variable_set :@variant_override, variant_override

variant_configurations = [
  [true, 20],
  [false, 20]
]
variant_override_configurations = [
  [nil, nil],
  [true, nil],
  [false, nil],
  [nil, 10],
  [true, 10],
  [false, 10]
]

results = [["VO OD", "VO OH", "V OD", "V OH", "OD", "OH"]]

variant_configurations.each do |variant_conf|
  set_stock(variant, *variant_conf)

  variant_override_configurations.each do |variant_override_conf|
    set_stock(variant_override, *variant_override_conf)

    results << [
      *variant_override_conf,
      *variant_conf,
      variant.on_demand,
      (variant.count_on_hand unless variant.on_demand)
    ]
  end
end

results.sort_by(&:to_s).each_with_index do |result, i|
  puts "|#{([i] + result).map { |value| " #{value.inspect} |".rjust(8) }.join}"
end
```

Reply:
@mkllnk Let's continue discussing this in #3067.